crates/compiler/mono/src/inc_dec.rs

```diff
@@ -1,1554 +1,1199 @@
-                        codegen,
-                        *symbol,
-                        (*expr).clone(),
-                        *layout,
-                        b,
-                        &b_live_vars,
-                    );
-
-                    b = pair.0;
-                    b_live_vars = pair.1;
-                }
-
-                return (b, b_live_vars);
+                new_let!(new_stmt)
             }
-        }
-
-        match stmt {
-            Let(symbol, expr, layout, cont) => {
-                let ctx = self.update_var_info(*symbol, layout, expr);
-                let (b, b_live_vars) = ctx.visit_stmt(codegen, cont);
-                ctx.visit_variable_declaration(
-                    codegen,
-                    *symbol,
-                    expr.clone(),
-                    *layout,
-                    b,
-                    &b_live_vars,
-                )
-            }
-
-            Join {
-                id: j,
-                parameters: _,
-                remainder: b,
-                body: v,
+            // Doesn't include higher order
+            CallType::LowLevel {
+                op: operator,
+                update_mode: _,
             } => {
-                // get the parameters with borrow signature
-                let xs = self.param_map.get_join_point(*j);
-
-                let (v, v_live_vars) = {
-                    let ctx = self.update_var_info_with_params(xs);
-                    ctx.visit_stmt(codegen, v)
-                };
-
-                let mut ctx = self.clone();
-                let v = ctx.add_dec_for_dead_params(xs, v, &v_live_vars);
-
-                update_jp_live_vars(*j, xs, v, &mut ctx.jp_live_vars);
-
-                let (b, b_live_vars) = ctx.visit_stmt(codegen, b);
-
-                (
-                    ctx.arena.alloc(Join {
-                        id: *j,
-                        parameters: xs,
-                        remainder: b,
-                        body: v,
-                    }),
-                    b_live_vars,
+                let borrow_signature = lowlevel_borrow_signature(arena, *operator);
+                let arguments_with_borrow_signature = arguments
+                    .iter()
+                    .copied()
+                    .zip(borrow_signature.iter().copied());
+                let (owned_arguments, borrowed_arguments) = arguments_with_borrow_signature
+                    .partition::<std::vec::Vec<_>, _>(
+                    |(_, ownership)| matches!(ownership, Ownership::Owned),
+                );
+
+                let new_stmt = dec_borrowed!(
+                    borrowed_arguments.iter().map(|(symbol, _)| symbol).copied(),
+                    stmt
+                );
+
+                let new_let = new_let!(new_stmt);
+
+                inc_owned!(
+                    owned_arguments.iter().map(|(symbol, _)| symbol).copied(),
+                    new_let
                 )
```

Comment:
this allocation of the two vectors is really not needed. Just an extra `.filter` should do

Reply:
in the `dec_borrowed` and `inc_owned` calls.